src/main/scala/karazin/scala/users/group/week2/homework/adt.scala

```diff
@@ -12,63 +15,61 @@ package karazin.scala.users.group.week2.homework
 object adt:
   
   enum ErrorOr[+V]:
+
+    case Some(x: V) extends ErrorOr[V]
+  
+    case SomeError(x: Throwable) extends ErrorOr[V]
+  
+    def flatMap[Q](f: V ⇒ ErrorOr[Q]): ErrorOr[Q] = 
+      this match
+        case ErrorOr.SomeError(q) ⇒ ErrorOr.SomeError(q)
+        case ErrorOr.Some(v) ⇒ try {
+          f(v)
+        } catch {
+          case NonFatal(e) ⇒ ErrorOr.SomeError(e)
+        }
     
-    // Added to make it compilable. Remove it.
-    case DummyCase
-    
-    /* 
-      Two case must be defined: 
-      * a case for a regular value
-      * a case for an error (it should contain an actual throwable)
-     */
   
-    /* 
-      The method is used for defining execution pipelines
-      Provide a type parameter, an argument and a result type
-      
-      Make sure that in case of failing the method with exception
-      no exception is thrown but the case for an error is returned
-    */ 
-    def flatMap = ???
-
-    /* 
-      The method is used for changing the internal object
-      Provide a type parameter, an argument and a result type
-      
-      Make sure that in case of failing the method with exception
-      no exception is thrown but the case for an error is returned
-     */
-    def map = ???
+    def map[Q](f: V ⇒ Q): ErrorOr[Q] =
+      this match
+        case ErrorOr.SomeError(t) ⇒ ErrorOr.SomeError(t)
+        case ErrorOr.Some(v) ⇒ try {
+          ErrorOr.Some(f(v))
+        } catch {
+          case NonFatal(e) ⇒ ErrorOr.SomeError(e)
+        }
   
-    /* 
-      The method is used for filtering
-      Provide a type parameter, an argument and a result type
-      
-      Make sure that in case of failing the method with exception
-      no exception is thrown but the case for an error is returned
-     */
-    def withFilter = ???
+  
+    def withFilter(p: V => Boolean): ErrorOr[V] = ???
   
     /* 
       The method is used for getting rid of internal box
       Provide a type parameter, an argument and a result type
     */
-    def flatten = ???
-    
+    def flatten[U](implicit ev: V <:< ErrorOr[U]): ErrorOr[U] =
```

Comment:
What does `V <:< ErrorOr[U]` mean?
Why do you use Scala 2 `implicit` instead of `given`.